packages/workers/src/block-indexer.ts

```diff
@@ -0,0 +1,173 @@
+          await this.waitForNewBlockToProcess(processingBlockNumber)
+        }
+
+        // if the indexer is too far behind, batch process blocks
+        const shouldBatch =
+          this._latestBlockNumber - processingBlockNumber >
+          BlockRunner._BATCH_SIZE
+
+        if (shouldBatch) {
+          const batchEndBlock = Math.min(
+            processingBlockNumber + BlockRunner._BATCH_SIZE,
+            this._latestBlockNumber,
+          )
+          const blockPromises = []
+          for (
+            let blockNum = processingBlockNumber;
+            blockNum < batchEndBlock;
+            blockNum++
+          ) {
+            blockPromises.push(
+              this._processBlockFn(blockNum)
+                .then(() => {
+                  logger.info(
+                    `[${this._chainName}] Processed block ${blockNum}`,
+                  )
+                })
+                .catch((error) => {
+                  logger.error(
+                    `[${this._chainName}] Error processing block ${blockNum}:`,
+                    error instanceof Error ? error.stack : String(error),
+                  )
+                  throw error
+                }),
+            )
+          }
+          await Promise.all(blockPromises)
+          processingBlockNumber = batchEndBlock
+        } else {
+          await this._processBlockFn(processingBlockNumber).catch((error) => {
+            logger.error(
+              `[${this._chainName}] Error processing block ${processingBlockNumber}:`,
+              error instanceof Error ? error.stack : String(error),
+            )
+            throw error
+          })
+
+          logger.info(
+            `[${this._chainName}] Processed block ${processingBlockNumber}`,
+          )
+          processingBlockNumber++
+        }
+
+        await logUpdate(this._chainId, this._provider)
+      } catch (error) {
+        logger.error(
+          `[${this._chainName}] Error processing block ${processingBlockNumber}:`,
+          error instanceof Error ? error.stack : String(error),
+        )
+
+        logger.info(
+          `curl -X POST --data '{"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x${processingBlockNumber.toString(
+            16,
+          )}", true],"id":1}' -H "Content-Type: application/json" ${
+            this._provider._getConnection().url
+          } | jq`,
+        )
+
+        retryCount++
+        if (retryCount >= BlockRunner._MAX_RETRIES) {
+          const earliestSafeBlock =
+            processingBlockNumber - BlockRunner._BATCH_SIZE
+
+          await this._onError(earliestSafeBlock)
+
+          logger.info(
+            `[${this._chainName}] Max retries exceeded for block ${processingBlockNumber}`,
+          )
+        }
+        await new Promise((res) => setTimeout(res, 5000 * retryCount))
+      }
+    }
+
+    async function logUpdate(chain: Chain, provider: CustomJsonRpcProvider) {
```

Comment:
```suggestion
    private async function logUpdate(chain: Chain, provider: CustomJsonRpcProvider) {
```

Should be private probably.